### PR TITLE
[CELEBORN-648][SPARK] Improve perf of SendBufferPool and logs about memory

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

- Replace index-based item access with an iterator for LinkedList.
- Always try to remove a buffer if SendBufferPool does not have a matched candidate, this change makes the total buffer number from `capacity+N-1` to `capacity` in worst cases.
- Some logs and code polish.

### Why are the changes needed?

Improve performance and logs, reduce memory consumption.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass GA.

### CODE REVIEW

File: client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java [hunk 1]
```diff
@@ -36,23 +37,23 @@ public static SendBufferPool get(int capacity) {
...
+    Iterator<byte[][]> iterator = buffers.iterator();
+    while (iterator.hasNext()) {
+      byte[][] candidate = iterator.next();
+      if (candidate.length == numPartitions) {
+        iterator.remove();
+        return candidate;
       }
     }
-    if (buffers.size() == capacity) {
-      buffers.removeFirst();
```
Comment (pan3793) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:53]: what's the benefit of shrink buffer during acquire? it should be ok to do it only in return

File: client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java [hunk 2]
```diff
@@ -223,10 +223,11 @@ public boolean insertRecord(
     if (getUsed() > pushSortMemoryThreshold
         && pageCursor + bytes8K > currentPage.getBaseOffset() + currentPage.size()) {
       logger.info(
-          "Memory Used across threshold, need to trigger push. Memory: "
-              + getUsed()
-              + ", currentPage size: "
-              + currentPage.size());
+          "partition {} memory used {} exceeds threshold {}, need to trigger push. currentPage size: {}",
```
[the remark this answers is not in the document]
Comment (pan3793) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java]: make sense, remove the partition info from the log

**Review by cfmcgrady — APPROVED** [review]

File: client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java (hunk 1 above)
Comment (waitinfuture) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:53]: The purpose is to make sure total number of buffers will not exceed capacity. Say we don't have matching buffer in the pool, and buffers.size() equals capacity, then acquireBuffer will return null, HashBasedShuffleWriter will new a new buffer, thus the total buffers will be ```capacity + 1```. For N tasks, total buffer number can be ```capacity + N```.

Comment (pan3793) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:53]: with this eager release, the worst case could be `capacity + N - 1`, right?

Comment (waitinfuture) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:53]: > with this eager release, the worst case could be `capacity + N - 1`, right?

I think you are right, and it's also not what we want. I think we should always try to remove if ```acquireBuffer``` returns no match.

Comment (pan3793) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:53]: >  I think we should always try to remove if acquireBuffer returns no match.

Make sense, changed

**Review by waitinfuture — APPROVED** [review]
LGTM except a nit, thanks!

File: client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java [hunk 3]
```diff
@@ -36,23 +37,27 @@ public static SendBufferPool get(int capacity) {
...
+    assert capacity > 0;
     this.capacity = capacity;
     buffers = new LinkedList<>();
   }
 
   public synchronized byte[][] acquireBuffer(int numPartitions) {
-    for (int i = 0; i < buffers.size(); i++) {
-      if (buffers.get(i).length == numPartitions) {
-        return buffers.remove(i);
+    Iterator<byte[][]> iterator = buffers.iterator();
```
[the remark this answers is not in the document]
Comment (pan3793) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:49]: typo perf

it's a linked list, not optimized for index based access

[the remark this answers is not in the document]
Comment (waitinfuture) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java:49]: I see, thanks.

**Comment by pan3793** [comment]
Thanks, merged to main/0.3

**Pull request closed without merging**